src/Servers/HttpSys/src/FeatureContext.cs

```diff
@@ -333,7 +333,18 @@ async Task<X509Certificate2> ITlsConnectionFeature.GetClientCertificateAsync(Can
         {
             if (IsNotInitialized(Fields.ClientCertificate))
             {
-                _clientCert = await Request.GetClientCertificateAsync(cancellationToken);
+                var method = _requestContext.Server.Options.ClientCertificateMethod;
+                if (method != ClientCertificateMethod.NoCertificate)
+                {
+                    // Check if a cert was already available on the connection.
+                    _clientCert = Request.ClientCertificate;
+                }
+
+                if (_clientCert == null && method == ClientCertificateMethod.AllowRenegotation)
+                {
+                    _clientCert = await Request.GetClientCertificateAsync(cancellationToken);
+                }
```

Comment:
Why did this method have to change if this PR is merely changing the default mode?

Does this mean that ITlsConnectionFeature.GetClientCertificateAsync() was broken before when `ClientCertificateMethod.AllowCertificate` was set? Would it previously allow renegotiation anyway? Should we backport this part of the change?

Reply:
In the 3.1 fix we were primarily addressing complaints about ClientCertificate triggering IO, and sync-over-async at that. There haven't been complaints about GetClientCertificateAsync, it's rarely used. I'm changing it now more for consistency and efficiency.